src/main/java/com/couchbase/lite/router/Router.java

```diff
@@ -1887,19 +1875,46 @@ private RevisionInternal update(Database _db, String docID, Body body, boolean d
 
         RevisionInternal result = null;
         try {
-            synchronized (_db) {
-                if (isLocalDoc) {
-                    result = _db.getStore().putLocalRevision(rev, prevRevID, true);
-                } else {
-                    result = _db.putRevision(rev, prevRevID, allowConflict);
+            if (isLocalDoc) {
+                // NOTE: putLocalRevision() does not use transaction internally with obeyMVCC=true
+
+                final Database fDb = _db;
+                final RevisionInternal _rev = rev;
+                final String _prevRevID = prevRevID;
+                final List<RevisionInternal> _revs = new ArrayList<RevisionInternal>();
+                try {
+                    fDb.getStore().runInTransaction(new TransactionalTask() {
+                        @Override
+                        public boolean run() {
+                            try {
+                                RevisionInternal r = fDb.getStore().putLocalRevision(_rev, _prevRevID, true);
+                                _revs.add(r);
+                                return true;
+                            } catch (CouchbaseLiteException e) {
+                                throw new RuntimeException(e);
```

Comment:
As [boolean run()](https://github.com/couchbase/couchbase-lite-java-core/blob/51b2bf2cccd0048fb2821cb527b315e3cc5e7102/src/main/java/com/couchbase/lite/TransactionalTask.java#L12) is not declared with throws Exception/CouchbaseLiteException. This is reason that re-throw exception with RuntimeException.